File: bindings/nodejs/examples/misc.js
'use strict';

// The remaining crypto_* modules, each small enough to share one file:
//
// - crypto_auth (Kupyna-KMAC): keyed message authentication.
// - crypto_kdf: deterministic subkey derivation from a master key.
// - crypto_generichash (Kupyna-256/512): one-shot and streaming hashing.
// - crypto_stream (Strumok-256): unauthenticated keystream cipher - no integrity, wrong key/
//   tampered ciphertext silently decrypts to different, wrong plaintext instead of throwing.
// - randombytes: CSPRNG-backed random bytes.
//
// Run: node examples/misc.js

const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

function authExample() {
  const key = dstu.authKeygen();
  const message = Buffer.from('a message both parties want to confirm is unmodified');
  const tag = dstu.auth(key, message);
  dstu.authVerify(key, message, tag);
  console.log('auth: tag verified');
}

function kdfExample() {
  const masterKey = dstu.kdfKeygen();
  const subkeyA = dstu.kdfDeriveSubkey(masterKey, 0, Buffer.from('encrypt_'));
  const subkeyB = dstu.kdfDeriveSubkey(masterKey, 1, Buffer.from('encrypt_'));
  assert.notDeepStrictEqual(subkeyA, subkeyB);
  console.log('kdf: subkey 0 and subkey 1 differ, as expected');
}

function generichashExample() {
  const oneShot = dstu.kupyna256(Buffer.from('hello world'));
  const hasher = new dstu.Kupyna256Hasher();
  hasher.update(Buffer.from('hello '));
  hasher.update(Buffer.from('world'));
  assert.deepStrictEqual(hasher.finalize(), oneShot);
  console.log(`generichash: kupyna256('hello world') = ${oneShot.toString('hex')}`);
}

function streamExample() {
  const key = dstu.streamKeygen();
  const sealed = dstu.streamEncrypt(key, Buffer.from('a message'));
  assert.deepStrictEqual(dstu.streamDecrypt(key, sealed), Buffer.from('a message'));
  console.log('stream: round-tripped (note: unauthenticated, no tamper detection)');
}

function randombytesExample() {
  const a = dstu.randombytesBuf(16);
  const b = dstu.randombytesBuf(16);
  assert.notDeepStrictEqual(a, b);
  console.log(`randombytes: two independent 16-byte draws, e.g. ${a.toString('hex')}`);
}

authExample();
kdfExample();
generichashExample();
streamExample();
randombytesExample();

File: bindings/cpp/examples/misc.cpp
// Smaller primitives that don't need their own dedicated example file: crypto_auth, crypto_kdf,
// crypto_generichash, crypto_stream, randombytes.

#include "dstu/dstu.hpp"

#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

namespace {
void Die(const std::string &what) {
  std::cerr << "error: " << what << "\n";
  std::exit(1);
}
std::vector<std::uint8_t> ToBytes(const std::string &s) { return std::vector<std::uint8_t>(s.begin(), s.end()); }

void DemoRandombytes() {
  auto buf = dstu::RandomBytes(16);
  std::cout << "RandomBytes: filled " << buf.size() << " random bytes\n";
}

void DemoAuth() {
  auto key = dstu::AuthKey::Generate();
  const std::string message = "a message both parties want to confirm is unmodified";
  auto tag = key.Compute(ToBytes(message));
  bool ok = true;
  try {
    key.Verify(ToBytes(message), tag);
  } catch (const dstu::CryptoError &) {
    ok = false;
  }
  std::cout << "crypto_auth: tag verified = " << (ok ? "true" : "false") << "\n";
}

void DemoKdf() {
  auto key = dstu::KdfMasterKey::Generate();
  auto encryptSubkey = key.DeriveSubkey(0, ToBytes("encrypt_"));
  auto macSubkey = key.DeriveSubkey(1, ToBytes("mac_key_"));
  (void)macSubkey;
  std::cout << "crypto_kdf: derived two distinct " << encryptSubkey.size() << "-byte subkeys from one master key\n";
}

void DemoGenerichash() {
  auto digest = dstu::GenericHash256(ToBytes("hello world"));
  std::cout << "crypto_generichash: Kupyna-256(\"hello world\") = " << std::hex << std::setfill('0');
  for (auto b : digest) {
    std::cout << std::setw(2) << static_cast<int>(b);
  }
  std::cout << std::dec << "\n";
}

void DemoStream() {
  auto key = dstu::StreamCipherKey::Generate();
  const std::string message = "message";
  auto sealed = key.Encrypt(ToBytes(message));
  auto plaintext = key.Decrypt(sealed);
  std::cout << "crypto_stream: round-tripped " << plaintext.size()
            << " bytes (confidentiality only, no authentication)\n";
}
}  // namespace

int main() {
  DemoRandombytes();
  DemoAuth();
  DemoKdf();
  DemoGenerichash();
  DemoStream();
  return 0;
}
